### PR TITLE
Disallow disabled entities in a bundle import #4471

### PR DESCRIPTION
**Pull request by agzam**
### Won't allow Bundle-imports with disabled entities

> **Epic** #
> Close  https://github.com/threatgrid/iroh/issues/4471
> Related #1020 

<a name="qa">[§](#qa)</a> QA
============================

1. Set property configuration for ctia.features.disabled, e.g. `ctia.features.disable asset,actor,sighting,asset-mapping,asset-properties`
2. Run the app
3. Go to the Swagger console

**Expected:** pre-filled map of `/ctia/bundle/import` path should not contain any keys related to disabled entities, no 'asset', 'asset_refs', 'sighting', etc.

4. Try to submit a bundle that contains key(s) pertaining to disabled entities [anyway]

**Expected:** POST should fail with a message about "disallowed-keys"

```
intern: Bundle imports do not allow submitting disabled entities.
public: one line description for public release notes.
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

81035bc9 extends tests
325a7ced PR feedback
b901b7aa adds a test
34eec646 fixing ns header
9acd4d7f removing entities in the schema

### CODE REVIEW

File: src/ctia/bundle/routes.clj [hunk 1]
```diff
@@ -67,9 +67,28 @@
     :read-casebook
     :list-casebooks})
 
+(defn- keyword->str [key] (str (if (namespace key) "/" "") (name key)))
```
Comment (ereteog) [inline src/ctia/bundle/routes.clj:70]: We should have only simple keywords in bundle.

File: src/ctia/bundle/routes.clj [hunk 2]
```diff
@@ -67,9 +67,28 @@
...
+
+(defn- entity->bundle-keys
+  "For given entity key returns corresponding keys that may be present in Bundle schema.
+  e.g. :asset => [:assets :asset_refs]"
+  [entity-key]
+  (let [{:keys [entity plural]} (get (entities/all-entities) entity-key)
+        kw->snake-case-str      (fn [kw] (-> kw keyword->str (str/replace #"-" "_")))]
```
Comment (ereteog) [inline src/ctia/bundle/routes.clj:77]: We should have only simple keywords in bundle, you can remove `keyword->str`.
```suggestion
        kw->snake-case-str      (fn [kw] (-> kw name (str/replace #"-" "_")))]
```

**Review by ereteog — COMMENTED** [review]
In our discussion I linked twice the bundle route, one was supposed to link to the bulk ;-).
I opened a separate issue, but it should reuse the same logic https://github.com/threatgrid/iroh/issues/4549

File: src/ctia/bundle/routes.clj [hunk 3]
```diff
@@ -67,9 +67,28 @@
...
+    [(-> plural kw->snake-case-str keyword)
+     (-> entity kw->snake-case-str (str "_refs") keyword)]))
+
+(s/defn prep-bundle-schema :- s/Any
+  [{{:keys [enabled?]} :FeaturesService} :- APIHandlerServices]
+  (let [to-remove (->> (entities/all-entities)
+                       keys
+                       (filter (comp not enabled?))
+                       (mapcat entity->bundle-keys))]
+    (apply st/dissoc NewBundle to-remove)))
```
Comment (ereteog) [inline src/ctia/bundle/routes.clj:87]: I would use the positive couterpart to (slightly) ease the reading:
```suggestion
(s/defn prep-bundle-schema :- s/Any
  [{{:keys [enabled?]} :FeaturesService} :- APIHandlerServices]
  (->> (entities/all-entities)
        keys
        (filter enabled?)
        (mapcat entity->bundle-keys)
        (apply st/dissoc NewBundle to-remove)))
```

File: src/ctia/bundle/routes.clj [hunk 4]
```diff
@@ -67,9 +67,28 @@
...
+(defn- entity->bundle-keys
+  "For given entity key returns corresponding keys that may be present in Bundle schema.
+  e.g. :asset => [:assets :asset_refs]"
+  [entity-key]
+  (let [{:keys [entity plural]} (get (entities/all-entities) entity-key)
+        kw->snake-case-str      (fn [kw] (-> kw keyword->str (str/replace #"-" "_")))]
+    [(-> plural kw->snake-case-str keyword)
+     (-> entity kw->snake-case-str (str "_refs") keyword)]))
+
+(s/defn prep-bundle-schema :- s/Any
```
Comment (ereteog) [inline src/ctia/bundle/routes.clj:81]: In addition of your route level test could you add a unit test for this function?
Testing at the most granular level has the advantage of:
- easing the debugging
- better cover edge cases (which are harder to combine at a higher level)

**Review by ereteog — APPROVED** [review]
LGTM (minor remarks)

File: src/ctia/bundle/routes.clj (hunk 2 above)
Comment (agzam) [reply src/ctia/bundle/routes.clj:77]: Initially when I added `keyword->str`, I thought maybe we can later move it into some `utils` namespace and re-use, but you're right, it's YAGNI. I'm gonna get rid of it.

File: src/ctia/bundle/routes.clj (hunk 4 above)
Comment (agzam) [reply src/ctia/bundle/routes.clj:79]: Take a look: 57fe0cafefb93d1d4b66a462698939e5fa9250d1

**Comment by hshahman** [comment]
Verified in rel-v1.64
